File: Dynamic Programming/Maximum Subarray .cs
/*Time: O(n)
Find the contiguous subarray within an array (containing at least one number) which has the largest sum.

For example, given the array [-2,1,-3,4,-1,2,1,-5,4],
the contiguous subarray [4,-1,2,1] has the largest sum = 6.
*/
public class Solution {
    public int MaxSubArray(int[] nums) {
        int sum = 0; int maxSum = 0; int nMax = int.MinValue;
        int neg=0;
        for(int i = 0; i<nums.Length; i++){
            sum+=nums[i];
            if(sum > 0) maxSum = Math.Max(maxSum , sum);
            else{
                neg++;
                nMax = Math.Max(nums[i], nMax);
                sum = 0;
            }
               
        }
        if(neg == nums.Length) return nMax;
        return Math.Max(sum , maxSum);
    }
}

File: MaximumSubArray.cpp
/*
Find the contiguous subarray within an array (containing at least one number) which has the largest sum.

For example, given the array [-2,1,-3,4,-1,2,1,-5,4],
the contiguous subarray [4,-1,2,1] has the largest sum = 6.
*/
class Solution {
public:
    int maxSubArray(int A[], int n) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        if(n==1) return A[0];
        int sum = 0, mx = 0,nmx = -99999999, cnt = 0;
        for(int i=0;i<n;i++){
            sum+=A[i];
            if(sum>mx) mx = sum;
            else if(sum<0) { cnt++; sum =0; nmx=max(nmx,A[i]);}
        }
        if(cnt == n) return nmx;
        return max(sum,mx);
    }
};